Implement DataAPI's refarray/refvalue/refpool interface

nalimilan opened the pull request:
It turns out it is much simpler to return a new special `CategoricalRefPool` type from `refpool` than to return directly the `CategoricalPool`. Indeed, `CategoricalPool` is not currently aware of whether its parent `CategoricalArray`  allows for missing values or not. Adding this as a type parameter requires adding a similar parameter to `CategoricalValue`, making it more complex for users than it needs to be.

Supersedes https://github.com/JuliaData/CategoricalArrays.jl/pull/297. Cc: @quinnj @bkamins

bkamins commented:
Thank you for working on this. The only problems are:
1. you do not define `similar` and `reshape` for this type (probably not a big deal)
2. `LinearIndices` works incorrectly with it (admittedly it is not very well documented in Julia Base). I think this should be fixed (have a look at OffsetArrays.jl - it uses `IdOffsetRange` type to work around this issue - maybe @timholy can comment more here)

However, maybe the simplest thing is not to make `CategoricalRefPool` an `AbstractVector` as DataAPI.jl does not require it, it only requires:
>  `refpool(A)[refarray(A)[I...]]` must be equal  to `A[I...]` invariant, which is weaker

nalimilan commented:
Yes, I've noticed the incorrect behavior of `LinearIndices`. It's not great, but `IdOffsetRange` is really to complex to be worth using here. We could also just throw an error from `LinearIndices` (it's not likely to be used for this type anyway).

I've just pushed a commit to stop subtyping `AbstractArray`, but it requires reimplementing more methods manually (and some are still missing, like `iterate`). Overall I'm not sure it's a good tradeoff.

DataAPI currently only requires `getindex` to be defined, but as noted at https://github.com/JuliaData/DataAPI.jl/pull/1/#issuecomment-626141115 we actually need more methods for `refpool` to be useful. In particular we need to be able to get the list of valid indices, otherwise we would have to go over the data once just to find out the possible levels. And Arrow.jl also relies on more functions, like `eltype` (https://github.com/JuliaData/Arrow.jl/pull/12#discussion_r483145212). So the docs need updatinh.

nalimilan commented:
Ah, I forgot to implement `keys`, `values` and `pairs` as you suggested at https://github.com/JuliaData/DataAPI.jl/pull/1. So I'm even more tempted to continue subtyping `AbstractArray` to get all of these for free...

bkamins commented:
Then let us keep subtyping and make `LinearIndices` throw an error (so we will learn if someone reports a problem with it and this can be fixed later).

For `similar` and `reshape` probably we then should throw an error (if we do not already - I have checked `similar` and I think it threw an error - but I might be wrong here). Again for the same reason - we can add it later if it is needed (probably not).

nalimilan commented:
OK, I've implemented `LinearIndices` as it's easy to do in accordance to what [the manual recommends](https://docs.julialang.org/en/v1/devdocs/offset-arrays/). I've also added tests that `reshape` and `similar` fail.

bkamins commented on src/array.jl:
```diff
@@ -1050,3 +1050,27 @@ StructTypes.construct(::Type{<:CategoricalArray{Union{Nothing, T}}},
+Base.size(x::CategoricalRefPool{T}) where {T} = (length(x.pool) + (T >: Missing),)
+Base.axes(x::CategoricalRefPool{T}) where {T} =
+    ((T >: Missing ? 0 : 1):length(x.pool),)
+Base.LinearIndices(x::CategoricalRefPool) = axes(x, 1)
```
:clap:

quinnj replied:
Can you explain what this is needed for? I'm not familiar with LinearIndices. Does it have to do with defining `axes` as well? Or why is that needed?

bkamins replied:
This is needed because by default `Base.LinearIndices` would return `1:length(x)` even if the valid indices are from `0` to `length(x)-1` incorrectly.

It is not likely that in normal use cases anyone would use `Base.LinearIndices` but who knows what users will try to do.

bkamins commented on src/array.jl:
```diff
@@ -1050,3 +1050,27 @@ StructTypes.construct(::Type{<:CategoricalArray{Union{Nothing, T}}},
+end
+
+Base.IndexStyle(::Type{<: CategoricalRefPool}) = Base.IndexLinear()
+@inline function Base.getindex(x::CategoricalRefPool, i::Int)
```
A minor thing - don't we need `@propagate_inbounds` here?

nalimilan replied:
Not AFAIK. It's only needed when `@inbounds` should be propagated to a function that is called from there. But here we use `@inbounds` explicitly.

bkamins commented on src/array.jl:
```diff
@@ -1050,3 +1050,27 @@ StructTypes.construct(::Type{<:CategoricalArray{Union{Nothing, T}}},
+Base.LinearIndices(x::CategoricalRefPool) = axes(x, 1)
+
+DataAPI.refarray(A::CatArrOrSub) = refs(A)
+@inline function DataAPI.refvalue(A::CatArrOrSub{T}, i::Integer) where T
```
same here

bkamins replied:
we typically separate `function` form by newlines above and below

bkamins commented:
OK - it looks good to merge

bkamins commented:
Also maybe bump version?

nalimilan commented:
I'll do that directly on master.
